### Step 1: Mapping reads
#rule bwa_map:
	#input:
		#"data/genome.fa",
		#"data/samples/A.fastq"
	#output:
		#"mapped_reads/A.bam"
	#shell:
		#"bwa mem {input} | samtools view -Sb - > {output}"

#Since the rule has multiple input files, Snakemake will concatenate them separated by a whitespace

### Step 2: Generalizing the read mapping rule

rule bwa_map:
	input:
		"data/genome.fa",
		"data/samples/{sample}.fastq" #wildcards
	output:
		"mapped_reads/{sample}.bam"
	shell:
		"bwa mem {input} | samtools view -Sb - > {output}"

### Step 3: Sorting read alignments

rule samtools_sort: 
	input:
		"mapped_reads/{sample}.bam"
	output:
		"sorted_reads/{sample}.bam"
	shell:
		"samtools sort {input} {output}"







